kube/src/service/auth.rs

```diff
@@ -0,0 +1,191 @@
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+
+use futures::{ready, Future};
+use http::{header::AUTHORIZATION, Request};
+use hyper::Body;
+use pin_project::pin_project;
+use tower::{layer::Layer, BoxError, Service};
+
+use crate::{config::Authentication, Result};
+
+/// `Layer` to decorate the request with `Authorization` header.
+pub struct AuthLayer {
+    auth: Authentication,
+}
+
+impl AuthLayer {
+    pub(crate) fn new(auth: Authentication) -> Self {
+        Self { auth }
+    }
+}
+
+impl<S> Layer<S> for AuthLayer
+where
+    S: Service<Request<Body>>,
+{
+    type Service = AuthService<S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        AuthService {
+            auth: self.auth.clone(),
+            service,
+        }
+    }
+}
+
+pub struct AuthService<S>
+where
+    S: Service<Request<Body>>,
+{
+    auth: Authentication,
+    service: S,
+}
+
+impl<S> Service<Request<Body>> for AuthService<S>
+where
+    S: Service<Request<Body>> + Clone,
+    S::Error: Into<BoxError>,
+{
+    type Error = BoxError;
+    type Future = AuthFuture<S>;
+    type Response = S::Response;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx).map_err(Into::into)
+    }
+
+    fn call(&mut self, mut req: Request<Body>) -> Self::Future {
+        // Comment from `AsyncFilter`
+        // > In case the inner service has state that's driven to readiness and
+        // > not tracked by clones (such as `Buffer`), pass the version we have
+        // > already called `poll_ready` on into the future, and leave its clone
+        // > behind.
+        let service = self.service.clone();
+        let service = std::mem::replace(&mut self.service, service);
```

Comment:
Looks like this is based on [AsyncFilter](https://docs.rs/tower/0.4.4/tower/filter/struct.AsyncFilter.html), so while I don't know if the the comment applies to how we use it (as a pre-step for the main service), at least it has an origin. It would be nice if tower had support for something like this out of the box. Doing an oauth dance isn't that uncommon.